keel-bakery-plugin/src/main/kotlin/com/netflix/spinnaker/keel/bakery/artifact/ImageHandler.kt

```diff
@@ -0,0 +1,167 @@
+import com.netflix.spinnaker.keel.actuation.ArtifactHandler
+import com.netflix.spinnaker.keel.api.actuation.Task
+import com.netflix.spinnaker.keel.api.actuation.TaskLauncher
+import com.netflix.spinnaker.keel.api.artifacts.DebianArtifact
+import com.netflix.spinnaker.keel.api.artifacts.DeliveryArtifact
+import com.netflix.spinnaker.keel.bakery.BaseImageCache
+import com.netflix.spinnaker.keel.clouddriver.ImageService
+import com.netflix.spinnaker.keel.core.NoKnownArtifactVersions
+import com.netflix.spinnaker.keel.events.ArtifactRegisteredEvent
+import com.netflix.spinnaker.keel.model.Job
+import com.netflix.spinnaker.keel.persistence.KeelRepository
+import com.netflix.spinnaker.keel.persistence.NoSuchArtifactException
+import com.netflix.spinnaker.keel.telemetry.ArtifactCheckSkipped
+import org.slf4j.LoggerFactory
+import org.springframework.context.ApplicationEventPublisher
+
+class ImageHandler(
+  private val repository: KeelRepository,
+  private val baseImageCache: BaseImageCache,
+  private val igorService: ArtifactService,
+  private val imageService: ImageService,
+  private val publisher: ApplicationEventPublisher,
+  private val taskLauncher: TaskLauncher,
+  private val defaultCredentials: BakeCredentials
+) : ArtifactHandler {
+
+  override suspend fun handle(artifact: DeliveryArtifact) {
+    if (artifact is DebianArtifact) {
+      if (taskLauncher.correlatedTasksRunning(artifact.correlationId)) {
+        publisher.publishEvent(
+          ArtifactCheckSkipped(artifact.type, artifact.name, "ActuationInProgress")
+        )
+      } else {
+        val latestVersion = artifact.findLatestVersion()
+        val latestBaseImageVersion = artifact.getLatestBaseImageVersion()
+        val image = imageService.getLatestImage(artifact.name, "test")
+
+        val imageMissing = image == null
+        val versionsDiffer = image?.appVersion != latestVersion || image?.baseAmiVersion != latestBaseImageVersion
+        val regionsDiffer = !(image?.regions ?: emptySet()).containsAll(artifact.vmOptions.regions)
+        if (imageMissing || versionsDiffer || regionsDiffer) {
+          launchBake(artifact, latestVersion, regionsDiffer)
+        }
+      }
+    }
+  }
+
+  private fun DebianArtifact.getLatestBaseImageVersion() =
+    baseImageCache.getBaseImage(vmOptions.baseOs, vmOptions.baseLabel)
+
+  /**
+   * First checks our repo, and if a version isn't found checks igor.
+   */
+  private suspend fun DebianArtifact.findLatestVersion(): String {
+    try {
+      val knownVersion = repository
+        .artifactVersions(this)
+        .firstOrNull()
+      if (knownVersion != null) {
+        log.debug("Latest known version of $name = $knownVersion")
+        return knownVersion
+      }
+    } catch (e: NoSuchArtifactException) {
+      log.debug("Latest known version of $name = null")
+      if (!repository.isRegistered(name, type)) {
+        // we clearly care about this artifact, let's register it.
+        repository.register(this)
```

Comment:
This could `throw` and prevent execution of the block below. Wrap it in a `try/catch`?

Reply:
why would we want to carry on trying to do anything if this throws an exception?